raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java

```diff
@@ -119,6 +122,93 @@ public void testNonMonotonicLocalEndOffsetUpdate() {
+        assertFalse(state.updateReplicaState(node1, ++fetchTime, new LogOffsetMetadata(10L), 11L));
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(localId).lastCaughtUpTimestamp(), currentTime);
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(node1).lastCaughtUpTimestamp(), caughtupTime);
+
+        // Node 1 catches up to leader
+        assertFalse(state.updateReplicaState(node1, ++fetchTime, new LogOffsetMetadata(11L), 11L));
+        caughtupTime = fetchTime;
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(localId).lastCaughtUpTimestamp(), currentTime);
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(node1).lastCaughtUpTimestamp(), caughtupTime);
+
+        // Node 1 falls behind
+        assertFalse(state.updateReplicaState(node1, ++fetchTime, new LogOffsetMetadata(50L), 100L));
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(localId).lastCaughtUpTimestamp(), currentTime);
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(node1).lastCaughtUpTimestamp(), caughtupTime);
+
+        // Node 1 catches up to the last fetch offset
+        int prevFetchTime = fetchTime;
+        assertFalse(state.updateReplicaState(node1, ++fetchTime, new LogOffsetMetadata(102L), 200L));
+        caughtupTime = prevFetchTime;
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(localId).lastCaughtUpTimestamp(), currentTime);
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(node1).lastCaughtUpTimestamp(), caughtupTime);
+
+        // Node2 has never caught up to leader
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(node2).lastCaughtUpTimestamp(), -1L);
+        assertTrue(state.updateReplicaState(node2, ++fetchTime, new LogOffsetMetadata(202L), 300L));
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(node2).lastCaughtUpTimestamp(), -1L);
+        assertFalse(state.updateReplicaState(node2, ++fetchTime, new LogOffsetMetadata(250L), 300L));
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(node2).lastCaughtUpTimestamp(), -1L);
+    }
+
+    @Test
+    public void testLastCaughtUpTimeObserver() {
+        int node1Index = 0;
+        int node1Id = 1;
+        int currentTime = 1000;
+        int fetchTime = 0;
+        int caughtupTime = -1;
+        LeaderState<?> state = newLeaderState(singleton(localId), 5L);
+        assertEquals(Optional.empty(), state.highWatermark());
+        assertEquals(emptySet(), state.nonAcknowledgingVoters());
+
+        // Node 1 falls behind
+        assertTrue(state.updateLocalState(++fetchTime, new LogOffsetMetadata(10L)));
+        assertFalse(state.updateReplicaState(node1Id, ++fetchTime, new LogOffsetMetadata(10L), 11L));
+        assertEquals(state.quorumResponseVoterStates(currentTime).get(localId).lastCaughtUpTimestamp(), currentTime);
```

Comment:
The expected value should always be the first argument. A few of these in this test class.